### ISSUE TITLE
`save` function errors when model has a secondary unique value

### ISSUE THREAD

**Issue by catdad** [balderdashy/sails-redis#92]
This is similar to https://github.com/balderdashy/sails-redis/issues/24, but seems to throw a different error. The gist is that if the Model contains a unique field other than the primary key, the record cannot be updated using the `save` function.

Repro code:

```javascript
var async = require('async');
var Waterline = require('waterline');
var waterline = new Waterline();

var Thing = Waterline.Collection.extend({
    tableName: 'thing',
    connection: 'default',
    attributes: {
        primary: {
            type: 'string',
            primaryKey: true,
            required: true
        },
        someId: {
            type: 'string',
            required: true,
            unique: true
        },
        data: 'string'
    },
    autoPK: false
});

waterline.loadCollection(Thing);

waterline.initialize({
    adapters: {
        redis: require('sails-redis')
    },
    connections: {
        default: {
            adapter: 'redis'
        }
    }
}, function(err, ontology) {
    function collection() {
        return ontology.collections.thing;
    }
    
    var theRecord;
    
    async.series([
        function deleteEverything(next) {
            collection().destroy({}).exec(next);
        },
        function createNew(next) {
            collection().create({
                primary: '1234',
                someId: '5678',
                data: 'llama'
            }).exec(function(err, record) {
                console.log('created', record);
                theRecord = record;
                next();
            });
        },
        function saveIsBroken(next) {
            theRecord.data = 'not llama';
            theRecord.save(next);
        },
        // function updateWorks(next) {
        //     collection().update({
        //         primary: '1234'
        //     }, {
        //         data: 'not llama'
        //     }).exec(function(err, updatedRecord) {
        //         console.log(updatedRecord);
        //         next();
        //     });
        // },
    ], function(err) {
        console.log(err);
        console.log('done');
        process.exit();
    });
});
```

The error being output:

```
Error (E_UNKNOWN) :: Encountered an unexpected error
AdapterError: Record does not satisfy unique constraints
    at C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\lib\database\index.js:787:41
    at C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\index.js:1170:13
    at try_callback (C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\index.js:592:9)
    at RedisClient.return_reply (C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\index.js:685:13)
    at ReplyParser.<anonymous> (C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\index.js:321:14)
    at emitOne (events.js:77:13)
    at ReplyParser.emit (events.js:169:7)
    at ReplyParser.send_reply (C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\lib\parser\javascript.js:300:10)
    at ReplyParser.execute (C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\lib\parser\javascript.js:211:22)
    at RedisClient.on_data (C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\index.js:547:27)
    at Socket.<anonymous> (C:\Users\kvatev\Desktop\sails-redis-update-bug\node_modules\sails-redis\node_modules\redis\index.js:102:14)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Socket.Readable.push (_stream_readable.js:110:10)
    at TCP.onread (net.js:523:20)

Details:  AdapterError: Record does not satisfy unique constraints
```

`save` works successfully if you remove the `unique` constraint on the `someId` field.

Further, as stated in #24, the Model `update` method works fine -- which I have included in the commented out `updateWorks` function.

**Comment by sailsbot** [comment]
Thanks for posting, @catdad.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by catdad** [comment]
This is a real bug with a small repro example, a clear error report, and plenty of info. It should not be closed. If no one is going to triage issues, then turn off the bot (or deprecate the package, add more contributors, etc.). Thanks.

**Comment by mikermcneil** [comment]
@catdad thanks for the heads up Ryan- the bug label will keep sailsbot at bay.

And sorry, didn't mean for sailsbot to run yet-- the next gen boatswain posts a comment asking for permission first, then waits seven days and if there hasn't been any activity, closes it. We're still working through the kinks while we get that online (cc @sgress454)

**Comment by fyodorvi** [comment]
This is a terrible issue, library is mostly unusable because of that.

**Comment by Ryanc1256** [comment]
@fyodorvi its actually fixed in my rebuild of this module but I'm waiting for someone to pull this or

https://github.com/balderdashy/waterline-criteria/pull/28

**Comment by fyodorvi** [comment]
@Ryanc1256 your fix doesn't seem to relate to this.

This problem is that when you call `.save()` method that enforces unique check doesn't respect that clashing keys are actually from the same record that you are updating.

**Comment by Ryanc1256** [comment]
@fyodorvi, no the thing I linked to is the thing I'm waiting on to get merged as it fixes a few things in my rebuild and my rebuild needs it to work fully.

**Comment by mjasnikovs** [comment]
Just note, this is still active problem in 2017.
When model have params defined with "unique: true". "update" function is producing this error.

**Comment by johnabrams7** [comment]
@mjasnikovs @Ryanc1256 @fyodorvi @catdad Hey everyone, we're currently moving all the redis issues to the main [Sails repo](https://github.com/balderdashy/sails/issues) - this is boost exposure to the entire community. We appreciate all the current updates and welcome any further developments - thanks!
